### DISCUSSION TITLE
Grouping update types "patch", "minor", "pin" and "digest" with enabled "git-submodules" manager behaves somehow wrong

### DISCUSSION THREAD

**Discussion by icepuma** [renovatebot/renovate#22098] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.79.1

### If you're self-hosting Renovate, select which platform you are using.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

Linux

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

When using the `"packageRules"` config

```json
{
	"updateTypes": ["patch", "minor", "pin", "digest"],
	"groupName": "non-major.versions"
}
```

and enable the `git-submodules` - manager

```js
"git-submodules": {
	enabled: true,
	groupName: "Git Submodules",
}
```

I see that the [PR description](https://github.com/open-source-issues/renovate-submodule/pull/8) is "flaky":
* `renovate` detects (in my case) changes for the GitHub actions and my submodule on one run creates the commit
* the next `renovate` run detects, that the GitHub actions are already up 2 date on the `renovate` branch and doesn't touch them, but detects a change for the submodule and creates a new commit with just my submodule digest
* the next `renovate` run detects, that GitHub actions need an update, because the PR only contains the submodule digest bump and creates a new commit containing all changes
* this repeats and repeats

I would expect the `git-submodules` - manager to detect, that the submodule is already bumped and does not report a new change every time `renovate` runs/

### What you tried so far.

I have a repo reproducing the behaviour https://github.com/open-source-issues/renovate-submodule

### Relevant debug logs

DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=true (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
DEBUG: Branch dep actions/checkout in .github/workflows/renovate.yml is already updated (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
**DEBUG: No content changed (repository=open-source-issues/renovate-submodule, packageFile=.github/workflows/renovate.yml, branch=test-renovate/non-major.versions)
       "depName": "actions/checkout"**
DEBUG: Branch dep renovatebot/github-action in .github/workflows/renovate.yml is already updated (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
**DEBUG: No content changed (repository=open-source-issues/renovate-submodule, packageFile=.github/workflows/renovate.yml, branch=test-renovate/non-major.versions)
       "depName": "renovatebot/github-action"**
 **INFO: Updating submodule renovate-submodule-include (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)**
DEBUG: Updated 1 package files (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
DEBUG: Updated 1 lock files (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
       "updatedArtifacts": ["renovate-submodule-include"]
DEBUG: 2 file(s) to commit (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
DEBUG: Preparing files for committing to branch test-renovate/non-major.versions (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
DEBUG: git commit (repository=open-source-issues/renovate-submodule, branch=test-renovate/non-major.versions)
       "deletedFiles": [],
       "ignoredFiles": [],
       "result": {
         "author": null,
         "branch": "test-renovate/non-major.versions",
         "commit": "42c5bf8bf62d1a4f3d19030956b31bda19d5e9c0",
         "root": false,
         **"summary": {"changes": 1, "insertions": 1, "deletions": 1}**
       }

**Comment by rarkins** [comment]
Does the problem only occur when you're grouping?

**Reply by icepuma** [reply]
When I use another package rules configuration e.g.

```js
packageRules: [
	{
		updateTypes: ["patch", "minor"],
		groupName: "non-major.versions",
	},
	{
		updateTypes: ["pin", "digest"],
		groupName: "pin and digest",
	},
],
```

the submodule update gets its own PR which is then fine

**Reply by rarkins** [reply]
Does the group PR contain _both_ a digest update and a patch or minor update for the same dependency?

**Reply by icepuma** [reply]
But am I correct, that the `git-submodules` - manager currently has no proper detection logic, whether a submodule bump already exists on a `renovate` PR?

**Reply by rarkins** [reply]
If you're using repositoryCache=enabled then you should see it remembers which branches have which updates. Otherwise, it needs to try to test the branch each run to see if it's got the updates expected, and if not then it will rebase and try again.

**Reply by icepuma** [reply]
> Does the group PR contain _both_ a digest update and a patch or minor update for the same dependency?

It pins to a digest and updates in the same PR (at least from my understanding):

![2023-05-11-172723_1087x349_scrot](https://github.com/renovatebot/renovate/assets/362487/80d40b22-c5af-49fe-89c7-3e242a21d317)

**Reply by icepuma** [reply]
> If you're using repositoryCache=enabled then you should see it remembers which branches have which updates. Otherwise, it needs to try to test the branch each run to see if it's got the updates expected, and if not then it will rebase and try again.

Okay, then it comes down to the `manager` itself if it would be able to detect such changes on their own, right?
Because I saw that `helm` and others are able to detect that nothing changed for them.

**Reply by rarkins** [reply, marked as answer]
My advice:
- Never combine pinning and non-pinning in the same PR (it has a high chance of failure)
- Avoid combining digest and non-digest in same PRs (maybe it's ok sometimes, but sometimes not)

**Reply by icepuma** [reply]
So, grouping like
```js
packageRules: [
	{
		updateTypes: ["patch", "minor"],
		groupName: "non-major.versions",
	},
	{
		updateTypes: ["pin", "digest"],
		groupName: "pin and digest",
	},
],
```

is always cool?

**Reply by rarkins** [reply]
The first group - definitely yes.
The second group - probably yes, but "pin" updates should already be grouped together with themselves by default so I recommend leaving them as-is and just grouping digest together separeately

**Reply by icepuma** [reply]
Should I somehow dig into the `git-submodule` code to see whether we're able to implement a `No changes found` feature or is it working as expected as it is right now, that it always finds a change?

**Reply by rarkins** [reply]
Feel free to dig in. Ideally each manager can reliably determine changes
